horovod/run/run.py

```diff
@@ -0,0 +1,436 @@
+            ))
+
+            ssh_successful_to_all_hosts = False
+    if not ssh_successful_to_all_hosts:
+        exit(1)
+    return True
+
+
+def _launch_task_servers(host_names, local_host_names, driver_addresses,
+                         num_hosts, tmout,
+                         key, ssh_port=None):
+    """
+    executes the task server and service client task for registration on the
+    hosts.
+    :param host_names: list of addresses. for example,
+        ['worker-0','worker-1']
+        ['10.11.11.11', '10.11.11.12']
+    :type host_names: list(string)
+    :param local_host_names: names that are resolved to one of the addresses
+    of local hosts interfaces. For example,
+        set(['localhost', '127.0.0.1'])
+    :type local_host_names: set
+    :param driver_addresses: map of interfaces and their address and port for
+    the service. For example:
+        {
+            'lo': [('127.0.0.1', 34588)],
+            'docker0': [('172.122.10.1', 34588)],
+            'eth0': [('11.111.33.73', 34588)]
+        }
+    :type driver_addresses: map
+    :param num_hosts:
+    :type num_hosts: int
+    :param tmout:
+    :type tmout: horovod.spark.util.timeout.Timeout
+    :return:
+    :rtype:
+    """
+
+    def _exec_command(command):
+        host_output = six.StringIO()
+        try:
+            exit_code = safe_shell_exec.execute(command,
+                                                stdout=host_output,
+                                                stderr=host_output)
+            if exit_code != 0:
+                print(
+                    "Launching horovodrun task function was not "
+                    "successful:\n{host_output}"
+                        .format(host_output=host_output.getvalue()))
+                os._exit(exit_code)
+        finally:
+            host_output.close()
+        return exit_code
+
+    if ssh_port:
+        ssh_port_arg = "-p {ssh_port}".format(ssh_port=ssh_port)
+    else:
+        ssh_port_arg = ""
+
+    remote_command_format = \
+        'ssh -o StrictHostKeyChecking=no {host} {ssh_port_arg} ' \
+        '\'{python} -m horovod.run.horovod_task_fn {index} ' \
+        '{driver_addresses} {num_hosts} {timeout} {key}\''
+
+    local_command_format = '{python} -m horovod.run.horovod_task_fn {index} ' \
+                           '{driver_addresses} {num_hosts} {timeout} {key}'
+
+    args_list = []
+    for index in range(len(host_names)):
+        host_name = host_names[index]
+        if host_name in local_host_names:
+            command = local_command_format.format(
+                python=sys.executable,
+                index=codec.dumps_base64(index),
+                driver_addresses=codec.dumps_base64(driver_addresses),
+                num_hosts=codec.dumps_base64(num_hosts),
+                timeout=codec.dumps_base64(tmout),
+                key=codec.dumps_base64(key)
+            )
+        else:
+            command = remote_command_format.format(
```

Comment:
I think it'd be better to in-line formats, otherwise, it's hard to see which args are required since it's visually far.

Reply:
fixed.